src/com/reco1l/osu/ui/SettingsFragment.kt

```diff
@@ -0,0 +1,392 @@
+
+
+    override fun onLoadView() {
+
+        sectionSelector = findViewById(R.id.section_selector)!!
+        sectionSelector.removeAllViews()
+
+        fun createSectionButton(text: String, icon: Int, section: Section) {
+
+            val button = TextView(ContextThemeWrapper(context, R.style.settings_tab_text))
+
+            button.layoutWidth = 200f.dp
+            button.text = text
+            button.drawableLeft = requireContext().getDrawable(icon)!!
+            button.drawableLeft!!.setTint(Color.WHITE)
+
+            button.setOnClickListener {
+
+                sectionSelector.forEach {
+                    if (it == button) {
+                        it.setBackgroundResource(R.drawable.rounded_rect)
+                        it.background.setTint(0xFF363653.toInt())
+                    } else {
+                        it.background = null
+                    }
+                }
+
+                this.section = section
+                setPreferencesFromResource(section.xml, null)
+            }
+
+            sectionSelector.addView(button)
+
+            if (this.section == section) {
+                button.callOnClick()
+            }
+        }
+
+
+        if (Multiplayer.isMultiplayer) {
+
+            fun createDivider(text: String) {
+
+                val divider = TextView(ContextThemeWrapper(context, R.style.settings_tab_divider))
+                divider.text = text
+
+                sectionSelector.addView(divider)
+            }
+
+            createDivider("Multiplayer")
+            createSectionButton("Player", R.drawable.person_24px, Section.Player)
+            createSectionButton("Room", R.drawable.groups_24px, Section.Room)
+            createDivider("Game")
+
+        }
+
+
+        if (!Multiplayer.isMultiplayer) {
+            createSectionButton("General", R.drawable.grid_view_24px, Section.General)
+        }
+
+        createSectionButton("Gameplay", R.drawable.videogame_asset_24px, Section.Gameplay)
+        createSectionButton("Graphics", R.drawable.display_settings_24px, Section.Graphics)
+        createSectionButton("Sound", R.drawable.headphones_24px, Section.Sounds)
+
+        if (!Multiplayer.isMultiplayer) {
+            createSectionButton("Library", R.drawable.library_music_24px, Section.Library)
+        }
+
+        createSectionButton("Input", R.drawable.trackpad_input_24px, Section.Input)
+        createSectionButton("Advanced", R.drawable.manufacturing_24px, Section.Advanced)
+
+
+        sectionSelector[0].topMargin = 32f.dp
+        sectionSelector[sectionSelector.childCount - 1].bottomMargin = 32f.dp
+
+        findViewById<View>(R.id.close)!!.setOnClickListener {
+            dismiss()
+        }
+    }
+
+
+    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) = Unit
+
+
+    // For whatever reason this is restricted API when it wasn't in previous SDKs.
+    @SuppressLint("RestrictedApi")
+    override fun onBindPreferences() = when(section) {
```

Comment:
This is one hell of a method. Maybe each section can be separated into its own method to make this method easier to parse mentally?

Reply:
That's overcomplicate things.

Reply:
If we are going to add more functionality to settings that require explicit handling, the complexity of this method will grow at a faster rate.

Reply:
No logic should be there.

> the complexity of this method will grow at a faster rate.

Before that we should take advantage of the listeners that `SharedPreferences` provides instead.

Reply:
Sure, but that will add yet another complexity to this already large PR, so let's move that across for the time being.